Disable path exist check for remote filesystems

MartinVerges opened the pull request:
##### SUMMARY
This fixes the incorrect
   "Unable to mount ip:/path as it does not exist"
message when mounting remote cephfs and nfs targets.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
mount

##### ADDITIONAL INFORMATION
Try to mount any remote filesystem like cephfs to a local directory.
The ansible script will fail due to the src does not exist.
<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
before:  error "Unable to mount ip:/path as it does not exist"
after: success
```

MartinVerges committed:
Add verify_src option to allow mounting of remote or special file systems.

This fixes the incorrect
   "Unable to mount ip:/path as it does not exist"
message when mounting remote filesystems.

Akasurde commented:
@MartinVerges Thanks for the fix. Could you please add a change log entry for this change ? Thanks.

MartinVerges commented:
@Akasurde If I would know where, then I could add a change log entry. sorry

Akasurde commented:
@MartinVerges No issue. You can add a changelog entry in [here](https://github.com/ansible/ansible/tree/devel/changelogs/fragments). Thanks.

patrakov commented:
Hard-coding the list of network filesystems does not sound right. At the very least, you forgot "cifs" :)

I would suggest applying the same logic as in systemd: check for the _netdev mount option, and use the same list of network filesystems as them: https://github.com/systemd/systemd/blob/a859abf062cef1511e4879c4ee39c6036ebeaec8/src/basic/stat-util.c#L195

(note that e.g. for glusterfs we would still rely on _netdev).

Or maybe we should add a new flag that would add _netdev on linux and skip the device verification?

OTOH, I understand that someone will immediately criticize my proposal as linux-only. I don't know how to respond.

patrakov commented:
Well, it is not only remote filesystems that need the path check disabled. Another case that comes to mind is tmpfs, or a separate instance of proc for a chroot, or anything else mentioned with "nodev" in /proc/filesystems on Linux. And don't forget about other systems, e.g. on OpenBSD the in-memory filesystem is called "mfs". I don't think we can enumerate all the exceptions, so maybe providing an explicit flag to opt out of the check, or maybe even *always* disabling the check, is the way to go.

MartinVerges commented:
I changed the patch to add a generic option that allows to configure the behavior. This way it should be generic and don't need to maintain a list of "special" FS types. In addition I added a changelog as requested.

Akasurde commented:
Adding the flag using argument to check if the filesystem is network fs or not is OK for me but then we can't backport this change to older releases. This keeps older versions broken. 
On the other hand, I like the idea of `systemd` handling the `fs` magic number.

Let me know your views.

MartinVerges commented:
I'm not sure that any systemd dependency would be a good choice. Even if I like systemd, there are people who don't.

The backport is indeed something I would prefer. How about just checking the _netdev presence to prevent the check? It would be more generic but won't help with special fs types, therefore not perfect as well.

Akasurde commented:
How about `_netdev` and an additional check for fs which does not support this check ?

MartinVerges commented:
My doubts are that I won't be able to make a 100% correct list of possible filesystems where the check fails. Therefore someone else will be in the same situation, unable to mount a correct fs using this module. That's why I changed the implementation from the first simple list to the option.

As I want a fix in the backports, how about backport a patch that just removes the buggy check and newer version provide as provided here?

patrakov commented:
My preference would be to just remove the check and let the actual mount fail. No extra arguments.

MartinVerges commented:
@Akasurde did you have time to think if the current solution fits or what I need to change?

patrakov commented:
I have looked again at the diff. What catches my attention now is the fact that nfs mounts were documented to work without any additional options. Therefore, I am even more confident now that the addition of `verify_src` option is a wrong solution. I will prepare an alternative pull request later today.

patrakov commented:
A bit of history investigation reveals that the check of src has been added only 3 months ago, as a part of https://github.com/ansible/ansible/issues/59183, so a straight revert of commit 72023d7 (this is what I have been proposing so far) would reintroduce the bug. `verify_src` would also reintroduce the bug in the case where `verify_src` is false. So my position now is that issue 59183 was fixed incorrectly, and I am working on a more proper fix.

Akasurde commented:
@patrakov I think removing this check altogether is a good option. Let me know if you are working on reverting https://github.com/ansible/ansible/commit/72023d7462e78635264fd12bfdb23894b4163cba. 

Thanks.

patrakov commented:
I am working on a new MR that would revert 72023d7 and then fix #59183 better.

patrakov commented:
See #65869 instead.

Akasurde commented:
superseded by https://github.com/ansible-collections/ansible.posix/pull/33

Akasurde commented:
@MartinVerges Thanks for the contribution. I merged @patrakov's fix with the test in https://github.com/ansible-collections/ansible.posix/pull/33.
